Bluetooth: Shell: Make `gatt notify` more general

alwa-nordic opened the pull request:
Make `gatt notify` take the source handle and the data to send.

alwa-nordic committed:
Bluetooth: Shell: Make `gatt notify` more general

Make `gatt notify` take the source handle and the data to send.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

theob-pro commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -1294,7 +1328,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(gatt_cmds,
 		      "unregister pre-predefined test service",
 		      cmd_unregister_test_svc, 1, 0),
-	SHELL_CMD_ARG(notify, NULL, "[data]", cmd_notify, 1, 1),
+	SHELL_CMD_ARG(notify, NULL, "<handle> <conn> <str>", cmd_notify, 3, 0),
```
The number of mandatory arguments should be 4 because it include the command name.

alwa-nordic replied:
```suggestion
	SHELL_CMD_ARG(notify, NULL, "<handle> <conn> <str>", cmd_notify, 4, 0),
```

alwa-nordic commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -1294,7 +1328,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(gatt_cmds,
 		      "unregister pre-predefined test service",
 		      cmd_unregister_test_svc, 1, 0),
-	SHELL_CMD_ARG(notify, NULL, "[data]", cmd_notify, 1, 1),
+	SHELL_CMD_ARG(notify, NULL, "<handle> <conn> <str>", cmd_notify, 4, 0),
```
```suggestion
	SHELL_CMD_ARG(notify, NULL, "<handle> <str>", cmd_notify, 3, 0),
```

jori-nordic commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -1294,7 +1328,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(gatt_cmds,
 		      "unregister pre-predefined test service",
 		      cmd_unregister_test_svc, 1, 0),
-	SHELL_CMD_ARG(notify, NULL, "[data]", cmd_notify, 1, 1),
+	SHELL_CMD_ARG(notify, NULL, "<handle> <str>", cmd_notify, 3, 0),
```
could we have an arg for different data types:
e.g. `gatt notify [handle] bin deadbeef1337` will write those hex values and `gatt notify [handle] str "hello world"` will write the "hello world" string.

jori-nordic replied:
or make two commands: notify-str and notify-bin. But I think it's gonna duplicate code a bit

alwa-nordic replied:
I would prefer if the shell handled this somehow. Maybe like \x"deadbeef1337"?

Thalley replied:
Sending hex or string will work the same for us if using `hex2bin`. The receiving side can then choose whether or not to display the data as a string or hex.

alwa-nordic replied:
Updated PR: The command now takes a hex value, like other commands like `gatt write`.

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +949,73 @@ static int cmd_unregister_test_svc(const struct shell *sh,
-static void notify_cb(struct bt_conn *conn, void *user_data)
+static uint8_t found_attr(const struct bt_gatt_attr *attr, uint16_t handle,
+			  void *user_data)
 {
-	const struct shell *sh = user_data;
+	const struct bt_gatt_attr **found = user_data;
 
-	shell_print(sh, "Nofication sent to conn %p", conn);
+	*found = attr;
+
+	return BT_GATT_ITER_STOP;
+}
+
+static const struct bt_gatt_attr *find_attr(uint16_t handle)
+{
+	const struct bt_gatt_attr *attr = NULL;
+
+	bt_gatt_foreach_attr(handle, handle, found_attr, &attr);
+
+	return attr;
```
Should we consider adding a `bt_gatt_find_by_handle` function?

alwa-nordic replied:
Yes, I think so. It's already implemented as the internal function `find_attr` in `gatt.c`.

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +949,73 @@ static int cmd_unregister_test_svc(const struct shell *sh,
+		shell_error(sh, "Handle: Not a hex number '%s'.", arg_handle);
+		return -EINVAL;
+	}
+	if (!IN_RANGE(handle, 1, 0xffff)) {
```
```suggestion
	if (!IN_RANGE(handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE)) {
```

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +949,73 @@ static int cmd_unregister_test_svc(const struct shell *sh,
-	params.func = notify_cb;
-	params.user_data = (void *)sh;
-	SET_CHAN_OPT_ANY(params);
+	data_len = strnlen(arg_data, UINT16_MAX + 1);
+	if (data_len > UINT16_MAX) {
+		shell_error(sh, "The argument is too big to fit in a notification.");
+		return -EINVAL;
+	}
```
Should we consider using `hex2bin` here instead? Or is your intention to only support string values?

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
 	}
 
-	memset(&params, 0, sizeof(params));
+	if ((arg_data_len/2) > CHAR_SIZE_MAX) {
```
```suggestion
	if ((arg_data_len / 2) > CHAR_SIZE_MAX) {
```

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
-		shell_error(sh, "No clients have enabled notifications for the vnd1_echo CCC.");
-		return -ENOEXEC;
-	}
+	char *arg_handle = argv[1];
+	char *arg_data = argv[2];
```
```suggestion
	const char *arg_handle = argv[1];
	const char *arg_data = argv[2];
```

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
+	if (err) {
+		shell_error(sh, "Handle '%s': Not a valid hex number.", arg_handle);
+		return -EINVAL;
+	}
+	if (!IN_RANGE(handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE)) {
```
```suggestion
	}
	
	if (!IN_RANGE(handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE)) {
```

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
+	if ((arg_data_len/2) > CHAR_SIZE_MAX) {
+		shell_error(sh, "Data: Size exceeds legal attribute size.");
+		return -EINVAL;
+	}
+	data_len = hex2bin(arg_data, arg_data_len, data, sizeof(data));
```
```suggestion
	}
	
	data_len = hex2bin(arg_data, arg_data_len, data, sizeof(data));
```

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
+	int err;
+	size_t data_len;
+	unsigned long handle;
+	static char data[CHAR_SIZE_MAX];
```
I am wondering if we could reuse the `gatt_write_buf`, as you cannot (currently anyhow) perform both the write and notify command in parallel, right?

Thalley reviewed (changes requested):
A few code style comments and a suggestion to reuse existing buffer

Thalley commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -946,43 +950,81 @@ static int cmd_unregister_test_svc(const struct shell *sh,
+	int err;
+	size_t data_len;
+	unsigned long handle;
+	static char data[CHAR_SIZE_MAX];
```
Also consider using `BT_ATT_MAX_ATTRIBUTE_LEN` instead of `CHAR_SIZE_MAX`

alwa-nordic committed:
Bluetooth: Shell: Remove ad-hoc `CHAR_SIZE_MAX`

Remove ad-hoc `CHAR_SIZE_MAX` in favor of `BT_ATT_MAX_ATTRIBUTE_LEN`.
The maximum size of an attribute is the same as the maximum size of a
characteristic.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Shell: Add missing include

`include/zephyr/shell/shell_string_conv.h` uses `bool` from `stdbool.h`,
but the include was missing.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>
